docker/resource_docker_service_funcs.go

```diff
@@ -0,0 +1,1038 @@
+func resourceDockerServiceExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*ProviderConfig).DockerClient
+	if client == nil {
+		return false, nil
+	}
+
+	apiService, err := fetchDockerService(d.Id(), d.Get("name").(string), client)
+	if err != nil {
+		return false, err
+	}
+	if apiService == nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func resourceDockerServiceCreate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+	client := meta.(*ProviderConfig).DockerClient
+
+	serviceSpec, err := createServiceSpec(d)
+	if err != nil {
+		return err
+	}
+
+	createOpts := dc.CreateServiceOptions{
+		ServiceSpec: serviceSpec,
+	}
+
+	if v, ok := d.GetOk("auth"); ok {
+		createOpts.Auth = authToServiceAuth(v.(map[string]interface{}))
+	} else {
+		createOpts.Auth = fromRegistryAuth(d.Get("image").(string), meta.(*ProviderConfig).AuthConfigs.Configs)
+	}
+
+	if v, ok := d.GetOk("update_config"); ok {
+		createOpts.UpdateConfig, _ = createUpdateOrRollbackConfig(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("rollback_config"); ok {
+		createOpts.RollbackConfig, _ = createUpdateOrRollbackConfig(v.([]interface{}))
+	}
+
+	service, err := client.CreateService(createOpts)
+	if err != nil {
+		return err
+	}
+
+	if v, ok := d.GetOk("converge_config"); ok {
+		convergeConfig := createConvergeConfig(v.([]interface{}))
+		log.Printf("[INFO] Waiting for Service '%s' to be created with timeout: %v", service.ID, convergeConfig.timeoutRaw)
+		timeout, _ := time.ParseDuration(convergeConfig.timeoutRaw)
+		stateConf := &resource.StateChangeConf{
+			Pending:    serviceCreatePendingStates,
+			Target:     []string{"running", "complete"},
+			Refresh:    resourceDockerServiceCreateRefreshFunc(service.ID, meta),
+			Timeout:    timeout,
+			MinTimeout: 5 * time.Second,
+			Delay:      7 * time.Second,
+		}
+
+		// Wait, catching any errors
+		_, err := stateConf.WaitForState()
+		if err != nil {
+			// the service will be deleted in case it cannot be converged
+			if deleteErr := deleteService(service.ID, d, client); deleteErr != nil {
+				return deleteErr
+			}
+			if strings.Contains(err.Error(), "timeout while waiting for state") {
+				return &DidNotConvergeError{ServiceID: service.ID, Timeout: convergeConfig.timeout}
+			}
+			return err
+		}
+	}
+
+	d.SetId(service.ID)
+	return resourceDockerServiceRead(d, meta)
+}
+
+func resourceDockerServiceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ProviderConfig).DockerClient
+
+	apiService, err := fetchDockerService(d.Id(), d.Get("name").(string), client)
+	if err != nil {
+		return err
+	}
+	if apiService == nil {
+		d.SetId("")
+		return nil
+	}
+
+	var service *swarm.Service
+	service, err = client.InspectService(apiService.ID)
+	if err != nil {
+		return fmt.Errorf("Error inspecting service %s: %s", apiService.ID, err)
+	}
+
+	d.Set("version", service.Version.Index)
+	d.Set("name", service.Spec.Name)
+
+	d.SetId(service.ID)
+
```

Comment:
The `READ` method here needs read all the possible values from the results of `InspectService` and update the state. Ideally all of the Service's attributes will be read and updated from the API.

Reply:
All of them are a lot ... https://docs.docker.com/engine/api/v1.32/#operation/ServiceInspect
And not all of them are probably set... I see only the `image` which will always be set. What do you think @catsby ? Or is there a general _mapping_ method for existing properties?

Reply:
Sorry for any confusion, I meant specifically anything Terraform offers in the schema we need to make sure we read and update the state. If we don't, Terraform will be unable to detect _drift_, meaning any changes made to the service out of band of Terraform. That way future runs of `terraform plan` will detect the change and prompt the user for resolution.